Add `isAutofocussed` and `isReadOnly` to `MoneyInput` and `MoneyField`

montezume opened the pull request:
Closes #330 

#### Summary

* Adds `isAutofocussed` prop to `<MoneyInput>` and `<MoneyField>`
* Adds `isReadOnly` prop  to `<MoneyInput>` and `<MoneyField>`
* `<MoneyInput>`: Wraps `SingleValue` of react-select with a label pointing at the currencyCode input, so that it can be targeted using RTL.

montezume commented on src/components/inputs/money-input/money-input.js:
```diff
@@ -14,12 +15,25 @@ import currencies from './currencies.json';
 
+const SingleValue = ({ id, ...props }) => (
+  <components.SingleValue {...props}>
+    <label htmlFor={id}>{props.children}</label>
```
wrap the value with a label pointing at react-select's input.

tdeekens commented on src/components/fields/money-field/money-field.spec.js:
```diff
@@ -65,7 +62,7 @@ const renderMoneyField = (props, options) =>
   const { getByLabelText } = renderMoneyField();
   expect(getByLabelText('Amount')).toBeInTheDocument();
-  expect(getByLabelText('Currency Code')).toBeInTheDocument();
+  expect(getByLabelText('EUR')).toBeInTheDocument();
```
💯

tdeekens commented on src/components/fields/money-field/money-field.spec.js:
```diff
@@ -115,13 +112,20 @@ it('should call onBlur when amount input loses focus', () => {
   expect(onBlur).toHaveBeenCalled();
 });
 
+it('should have focus the amount input automatically when isAutofocussed is passed', () => {
```
should have focussed? or should focus the?

montezume replied:
🙈

tdeekens commented on src/components/inputs/money-input/money-input.js:
```diff
@@ -328,7 +349,9 @@ class MoneyInput extends React.Component {
-    onChange: PropTypes.func.isRequired,
+    isReadOnly: PropTypes.bool,
+    isAutofocussed: PropTypes.bool,
+    onChange: requiredIf(PropTypes.func, props => !props.isReadOnly),
```
Curious: is this a general pattern across other inputs or what we strive for and should change elsewhere to if not already?

montezume replied:
it's the same in `LocalizedTextInput`, and quite a few inputs have quite a bit of copy / paste from that input

tdeekens commented on src/components/inputs/money-input/money-input.js:
```diff
@@ -502,12 +527,18 @@ class MoneyInput extends React.Component {
             value={option}
-            isDisabled={this.props.isDisabled || hasNoCurrencies}
+            isDisabled={
+              this.props.isDisabled || hasNoCurrencies || this.props.isReadOnly
```
Should the select rather also be in readonly mode if this component is in it? Feels a bit odd to mix states in that readonly in one becomes disabled on another.

montezume replied:
@tdeekens react-select doesn't seem to have a `readonly` prop [ref](https://react-select.com/props)

montezume commented:
Open question:

I'm considering adding a prop, something like (isCurrencySelectionDisabled), where under this case, instead of rendering a react-select, we will render a normal label with it's for pointing at the amountInputId. That way, the user can click on `EUR`, and focus on the input-able input. Same as with LocalizedTextInput. This would be helpful for #339. 

Any comments @dferber90 @tdeekens

dferber90 commented:
This already happens (except for the label part) when you pass no currencies to `MoneyInput`. It might make sense to also disable the selection it in case only one currency is passed.

I'm undecided about the API. Adding a separate prop opens up new combinations which we'd have to give answers for. What would happen for:
- `isCurrencySelectionDisabled: false` and `currencies: []`
- `isCurrencySelectionDisabled: false` and `currencies: ['EUR']`
- `isCurrencySelectionDisabled: false` and `currencies: undefined`
- `isCurrencySelectionDisabled: true` and `currencies: ['EUR']`

It seems easier to say that "currency selection will be enabled when more than one currency is passed".

I don't see users wanting to dynamically switch between having currency selection enabled and disabled. It seems more like something that's decided while building a feature. So devs can then decide to omit `currencies`. Even if it's needed dynamically they could do `currencies={isCurrencySelectionDisabled ? [] : ['EUR', 'USD']}`.

Let's have a chat about it 📞

montezume commented:
Okay in [1de572c](https://github.com/commercetools/ui-kit/pull/362/commits/1de572c0b0f772accfdfb181f9e4d2e6dd4fc38d) I removed `isCurrencySelectDisabled`, and reworked the component to display a label with an htmlFor pointing at the amount input id if there are no currencies.

montezume commented:
I found too that the current docs are better represented by this new change, than by the current state in master 😆 
```
List of possible currencies. When not provided or empty, the component renders a label with the value's currency instead of a dropdown```
